eth/beacon/state_machines/base.py

```diff
@@ -83,25 +83,30 @@ def get_block_class(cls) -> Type[BaseBeaconBlock]:
             return cls.block_class
 
     @classmethod
-    @functools.lru_cache(maxsize=32)
     @to_tuple
-    def get_prev_blocks_to_slot(cls,
-                                last_block_hash: Hash32,
-                                chaindb: BaseBeaconChainDB,
-                                depth: int,
-                                start_slot: int) -> Iterable[BaseBeaconBlock]:
+    def get_prev_blocks(cls,
+                        last_block_hash: Hash32,
+                        chaindb: BaseBeaconChainDB,
+                        searching_depth: int,
+                        target_slot_number: int) -> Iterable[BaseBeaconBlock]:
+        """
+        Returns the previous blocks.
+
+        Since the slot numbers might be not continuous. The searching depth is ``searching_depth``
+        and the target slot depth is ``target_slot_number``.
+        """
         if last_block_hash == GENESIS_PARENT_HASH:
             return
 
         block = chaindb.get_block_by_hash(last_block_hash)
 
-        for _ in range(depth):
+        for _ in range(searching_depth):
             yield block
             try:
                 block = chaindb.get_block_by_hash(block.parent_hash)
             except (IndexError, BlockNotFound):
                 break
-            if block.slot_number < start_slot:
+            if block.slot_number < target_slot_number:
```

Comment:
I find it interesting and not intuitive that this search will include blocks which have a slot number lower than `target_slot_number`.  I assume that is part of the spec, but it might be something worth leaving an inline comment to be sure future readers of this code know that behavior is intentional.

Reply:
> I find it interesting and not intuitive that this search will include blocks which have a slot number lower than target_slot_number

Ohh, my goal is making this search return the blocks that **are greater than or equal to** `target_slot_number` (`min_slot_number`). It should return at least one block (the block that produced `backup_active_state`).

Long story version:

1. Startup: Setup empty states
2. Apply the genesis block (`B0`) and get the genesis states. Persist these states (`C0` and `A0`)
	* `recent_block_hashes: [ZERO_HASH32]  * (CYCLE_LENGTH * 2)`
4. Apply block `B1`, update active state (`A1`)
    ```python
    recent_block_hashes = get_new_recent_block_hashes(
            old_block_hashes=recent_block_hashes,
            parent_slot=B0.slot_number,
            current_slot=B1.slot_number,
            parent_hash=B0.hash
    )
    ```
	* `recent_block_hashes: [ZERO_HASH32] * (CYCLE_LENGTH * 2 - 1) + [B0.hash]`
5. Apply block `B2`, update active state (`A2`)
    ```python
    recent_block_hashes = get_new_recent_block_hashes(
            old_block_hashes=recent_block_hashes,
            parent_slot=B1.slot_number,
            current_slot=B2.slot_number,
            parent_hash=B1.hash
    )
    ```
	* `recent_block_hashes: [ZERO_HASH32] * (CYCLE_LENGTH * 2 - 2) + [B0.hash] + [B1.hash]`
6. Apply block `B3`, update active state (`A3`)
…
7. Apply block `B64`, update active state (`A64`) and crystallized state (`C64`), and also persist states.


Assume that we have the backup active state `A0` and want to create `B3’` (!= `B3`), so we have to reproduce `A2`. It requires the blocks starting from **the parent** of  “the first block after `backup_active_state_slot`”. This parent block should be the block that produced `A0`, so I set `min_slot_number=backup_active_state_slot`.

Sorry, it's long. Do you think it makes sense?

Reply:
Just noting that I was wrong in my first read of the code and yes, this does indeed only return down to block **greater than or equal to** `target_slot_number`